toucan_connectors/__init__.py

```diff
@@ -1,50 +1,55 @@
-with suppress(ImportError):
-    from .mongo.mongo_connector import MongoConnector
-with suppress(ImportError):
-    from .mssql.mssql_connector import MSSQLConnector
-with suppress(ImportError):
-    from .mysql.mysql_connector import MySQLConnector
-with suppress(ImportError):
-    from .odata.odata_connector import ODataConnector
-with suppress(ImportError):
-    from .oracle_sql.oracle_sql_connector import OracleSQLConnector
-with suppress(ImportError):
-    from .postgres.postgresql_connector import PostgresConnector
-with suppress(ImportError):
-    from .sap_hana.sap_hana_connector import SapHanaConnector
-with suppress(ImportError):
-    from .snowflake.snowflake_connector import SnowflakeConnector
-with suppress(ImportError):
-    from .toucan_toco.toucan_toco_connector import ToucanTocoConnector
-with suppress(ImportError):
-    from .toucan_toco.trello_connector import TrelloConnector
+ALL_CONNECTORS_MAPPING = [
+    ('adobe_analytics.adobe_analytics_connector', 'AdobeAnalyticsConnector'),
+    ('azure_mssql.azure_mssql_connector','AzureMSSQLConnector'),
+    ('dataiku.dataiku_connector', 'DataikuConnector'),
+    ('elasticsearch.elasticsearch_connector', 'ElasticsearchConnector'),
+    ('google_analytics.google_analytics_connector', 'GoogleAnalyticsConnector'),
+    ('google_big_query.google_big_query_connector', 'GoogleBigQueryConnector'),
+    ('google_cloud_mysql.google_cloud_mysql_connector', 'GoogleCloudMySQLConnector'),
+    ('google_my_business.google_my_business_connector', 'GoogleMyBusinessConnector'),
+    ('google_spreadsheet.google_spreadsheet_connector', 'GoogleSpreadsheetConnector'),
+    ('hive.hive_connector', 'HiveConnector'),
+    ('http_api.http_api_connector', 'HttpAPIConnector'),
+    ('micro_strategy.micro_strategy_connector', 'MicroStrategyConnector'),
+    ('mongo.mongo_connector', 'MongoConnector'),
+    ('mssql.mssql_connector', 'MSSQLConnector'),
+    ('mysql.mysql_connector', 'MySQLConnector'),
+    ('odata.odata_connector', 'ODataConnector'),
+    ('oracle_sql.oracle_sql_connector', 'OracleSQLConnector'),
+    ('postgres.postgresql_connector', 'PostgresConnector'),
+    ('sap_hana.sap_hana_connector', 'SapHanaConnector'),
+    ('snowflake.snowflake_connector', 'SnowflakeConnector'),
+    ('toucan_toco.toucan_toco_connector', 'ToucanTocoConnector'),
+    ('trello.trello_connector', 'TrelloConnector'),
+]
+HERE = os.path.dirname(__file__)
+
+
+def _import_connector(con_path, con_class):
+    try:
+        mod = import_module(f'.{con_path}', 'toucan_connectors')
+        con_type = getattr(mod, con_class).type
+    except ImportError:
+        # Retrieve the connector type by parsing the file
+        con_file_path = f'{HERE}/{con_path.replace(".", "/")}.py'
+        with open(con_file_path) as f:
+            con_code = f.read()
+        con_type = re.search(con_class + r'\(ToucanConnector\):.*type = [\'"](\w+)[\'"]', con_code, re.S).group(1)
```

Comment:
It's short and sweet with a regex, but I'm worried it's too brittle and covering all the cases would make the regex too complex (here for example we are not catching `type='truc'` (no spaces) or different order of attributes etc...

In this case I don't think that it would be overkill to use ast : 


```
import ast  # std lib
from astsearch import ASTPatternFinder, prepare_pattern  # requires pip install astsearch

connector_pattern = prepare_pattern('class ??(ToucanConnector): ??')
type_pattern = prepare_pattern('type=?')

def first_finder(_ast, pattern):
    return next(ASTPatternFinder(pattern).scan_ast(_ast))

with open('hive.py') as f:
    module = ast.parse(f.read())

connector = first_finder(module, connector_pattern)
print(first_finder(connector, type_pattern).value.s)

>> Hive

```

Reply:
+1 for using `ast` in a way or another (either directly from stdlib, or using [astroid](https://astroid.readthedocs.io/en/latest/) or [astsearch](https://pypi.org/project/astsearch/))

Reply:
Trying to put thoughts in this... we shouldn't have to use such "advanced" techniques for such a simple usage. Why don't we simply register the name (the `type`) of the connector in the `__init__` ? 

I think we were lying in the middle of a bridge between :

- an "automatic plugin" world where connectors are isolated and autonomous pieces of code
  that would "declare" themselves, The 2 big advantages of this approach being that all
  information is clearly encapsulated in the connector and that it could even be registered
  from outside this package;

- a "controlled" catalogue of connectors where a main module / function would statically know
  where to find plugins, try to load and register them.

It looks like we've implemend a strange mix of those 2 approaches but therefore without getting all the benefits. If we want to maintain an explicit catalogue in order to know which connector is available, then let's do that, and simply do something like (pseudo-code):

```python

CONNECTORS_CATALOGUE = {
  'hive': 'toucan_connectors.hive.HiveConnector`,
  'graphql: 'toucan_connectors.graphql.GraphqlConnector',
  # ...
}

AVAILABLE_CONNECTORS = {}

for connector_type, connector_module in CONNECTORS_CATALOGUE.items():
    try:
        connector = __import__(connector_module)
    except ImportError:
        logging.info('conector %s not available', connector_type)
    else:
        AVAILABLE_CONNECTORS[connector_type] = conncetor
```

What do you think?

Reply:
yes, sounds good. 

I think we should still add a check that a key of `CONNECTORS_CATALOGUE` is indeed the values of the the attribute `ToucanConnector.type` in the relevant class. This `type` string is going to be written in two places (`CONNECTORS_CATALOGUE` and `ToucanConnector.type`) and it is easy to get confused or make a typo.

Reply:
> it is easy to get confused or make a typo.

Indeed and I would then argue to remove the latter :)

Reply:
> Indeed and I would then argue to remove the latter :)

remove `ToucanConnector.type` ? 

I would indeed prefer a solution where there is only one declaration.